[MRG+1] add Convergence warning in LabelPropagation

musically-ut opened the pull request:
Otherwise, it remains unclear whether the convergence was reached or whether the algorithm ran out of iterations. Currently, all the test cases trigger this warning. That is what triggered the investigation which led to #5774.

musically-ut committed:
EHN: Show a Convergence warning if the max_iters were performed.

TomDLT commented:
LGTM

Can you adapt the tests in order not to raise any convergence warning (or to silence them with `ignore_warnings` if necessary)?

musically-ut commented:
Will do.

musically-ut commented:
I can add `with ignore_warnings():` in the tests, but it will still trigger the warnings in the doc-tests. I think I'll work on a more comprehensive solution which addresses the bug in the algorithm #5774 and this convergence issue together.

musically-ut commented:
I will re-base this after #9239 is merged in.

musically-ut committed:
DOC Indicate that max_iter is an integer.

musically-ut committed:
FIX Change max_iter in Doctests to 1000.

Otherwise, the Doctests resulted in ConvergenceWarning.

musically-ut commented:
I've rebased this against the current master and have silenced the warnings in the doctests.

jnothman commented on sklearn/semi_supervised/label_propagation.py:
```diff
@@ -32,7 +32,7 @@
 >>> from sklearn import datasets
 >>> from sklearn.semi_supervised import LabelPropagation
->>> label_prop_model = LabelPropagation()
+>>> label_prop_model = LabelPropagation(max_iter=1000)
```
Is the default max_iter something we should by changing while we are making other backwards incompatible changes? Or is the current default reasonable

musically-ut replied:
To be honest, I don't know. The `ConvergenceWarning` is partly there to allow users to adjust `max_iter` depending on their problem.

jnothman reviewed (commented):
Please add a test

musically-ut committed:
Add test for ConvergenceWarning.

musically-ut commented:
Added a test.

jnothman commented:
well it seems like changing a parameter from 30 to 1000 in order to be
assured convergence on a small dataset is a bit extreme...

On 5 Jul 2017 4:58 pm, "Utkarsh Upadhyay" <notifications@github.com> wrote:

> Added a test.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/5893#issuecomment-313018177>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz656MQE5LTQEyMNdLYUHygOCfOpC9ks5sKzQVgaJpZM4GmPUj>
> .
>

jnothman commented:
I'm not sure what makes something slow to converge here, except that for
spreading, iterations to convergrnce is indirectly proportional to alpha,
is related to number of samples but with a small coefficient, and is
unaffected by the proportion of the data unlabelled.

For propagation it seems to have a supralinear relation to proportion
unlabelled (with high variance), at least on iris.

A default of 30 seems reasonable for LabelSpreading in general. For the
propagation example at least, we should limit unlabelled to 10% of the data.



On 5 Jul 2017 5:55 pm, "Joel Nothman" <joel.nothman@gmail.com> wrote:

> well it seems like changing a parameter from 30 to 1000 in order to be
> assured convergence on a small dataset is a bit extreme...
>
> On 5 Jul 2017 4:58 pm, "Utkarsh Upadhyay" <notifications@github.com>
> wrote:
>
>> Added a test.
>>
>> —
>> You are receiving this because you commented.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/scikit-learn/scikit-learn/pull/5893#issuecomment-313018177>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AAEz656MQE5LTQEyMNdLYUHygOCfOpC9ks5sKzQVgaJpZM4GmPUj>
>> .
>>
>

jnothman commented:
See code at https://gist.github.com/jnothman/c70667210c2a7cf2793102c5ac177f7d

musically-ut commented:
Hmm. I'll look into it later today/tomorrow.

jnothman reviewed (commented):
Can you please also assert_no_warning in the convergence case?

musically-ut committed:
Change defaults for LabelPropagation and LabelSpreading

Also add assert_no_warnings.

musically-ut commented:
`LabelPropagation` and `LabelSpreading` do seem to have very different behavior when it comes to convergence. I've changed the default number of iterations for them (`1000` for `LabelPropagation` and 30 for `LabelSpreading`). I've also limited the number of unlabelled entries in the Doctests. I am tempted to set the `seed` in the doctest to make sure that we don't run into accidental failures. What do you think?

I've added `assert_no_warning` to one of the tests where convergence was being tested. Should I add it to others as well?

jnothman commented:
sounds good

On 21 Jul 2017 8:21 am, "Utkarsh Upadhyay" <notifications@github.com> wrote:

> LabelPropagation and LabelSpreading do seem to have very different
> behavior when it comes to convergence. I've changed the default number of
> iterations for them (1000 for LabelPropagation and 30 for LabelSpreading).
> I've also limited the number of unlabelled entries in the Doctests. I am
> tempted to set the seed in the doctest to make sure that we don't run
> into accidental failures. What do you think?
>
> I've added assert_no_warning to one of the tests where convergence was
> being tested. Should I add it to others as well?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/5893#issuecomment-316846666>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz65-TMpIRP25gO9bGEJkwJfgfo1Zoks5sP9LRgaJpZM4GmPUj>
> .
>

musically-ut commented:
> sounds good

So was that a "yes" to both:

 - Adding a seed to the Doctests
 - Adding `assert_no_warnings` to _all_ calls to `mdl.fit` in the code

?

jnothman commented:
I don't know that it needs to be on all calls to fit, but for those that
depend on convergence, why not?

On 21 Jul 2017 8:43 am, "Utkarsh Upadhyay" <notifications@github.com> wrote:

> sounds good
>
> So was that a "yes" to both:
>
>    - Adding a seed to the Doctests
>    - Adding assert_no_warnings to *all* calls to mdl.fit in the code
>
> ?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/5893#issuecomment-316851025>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz65K0iFVWQD2pfzWcgH25JJ4g6aoJks5sP9gggaJpZM4GmPUj>
> .
>

musically-ut committed:
Add seed for tests.

musically-ut committed:
Add assert_no_warn on all tests which rely on convergance.

musically-ut commented:
I've set `np.random.seed` for the tests. I am not sure how much of state is kept while switching from one set of tests to another and whether my setting the seed _globally_ may reduce the variance in the tests happening downstream. If that is an issue, I can create a test specific `RandomState` for each doctest.

Also, I've added `assert_no_warnings` on the tests which depend on convergence.

jnothman commented:
Don't set the seed globally. It is not safe for multi-threaded testing, if nothing else. But usually we'd just use an integer for `random_state`

jnothman reviewed (commented):
Now that you've put all the `assert_no_warning`s in there, I think they just add confusion. Sorry. If you want to test convergence, test it with `n_iter_`. If you want to test that no warnings are issued in convergence (and you should), add this assertion to `test_convergence_warning`.

jnothman commented:
Thanks

musically-ut committed:
Use a RandomState instead of setting seed globally

musically-ut committed:
Move assert_no_warnings to the dedicated test.

musically-ut commented:
Ready for review (not sure whether pushing commits to the branch sends out notifications on GitHub).

jnothman commented on sklearn/semi_supervised/label_propagation.py:
```diff
@@ -376,7 +383,7 @@ class LabelPropagation(BaseLabelPropagation):
 
     def __init__(self, kernel='rbf', gamma=20, n_neighbors=7,
-                 alpha=None, max_iter=30, tol=1e-3, n_jobs=1):
+                 alpha=None, max_iter=1000, tol=1e-3, n_jobs=1):
```
We can only make this change if we include it in the 0.19 final release. Just to note...

musically-ut replied:
I am not sure I follow so I'll rephrase what I understood: 

> If we create a separate PR just with this change, it can get included in the 0.19 final release.

Did I understand the note correctly? If so, I'll create one presently.

jnothman replied:
Yes :)

jnothman commented on sklearn/semi_supervised/label_propagation.py:
```diff
@@ -287,6 +288,12 @@ def fit(self, X, y):
                     alpha, self.label_distributions_) + y_static
             remaining_iter -= 1
 
+        if remaining_iter <= 1:
```
I think this is right given the current implementation, but the current implementation is buggy in that it seems you can never obtain n_iter_==max_iter.

Please fix the convergence condition and inline the code for _not_converged. I would then write something like:

```python
while self.n_iter_ < self.max_iter:
    if ... < tol:
        break
    Do stuff
    self.n_iter_ += 1
else:
    Warn
```

musically-ut replied:
Done.

jnothman reviewed (commented):
Yes, we see commits as they come, but it is often helpful to confirm what they mean with a comment.

Sorry to add more work. You can make the change of defaults a separate PR to make sure we get it into 0.19.

Also, please add your name to the list of file authors.

ogrisel commented:
Can you please rebase this on top of current master?

musically-ut commented:
Yes, will do.

Also, I'll change the definition for convergence as recommended.

musically-ut committed:
Fix bug with max_iter.

musically-ut committed:
Add myself to file authors.

musically-ut committed:
Fix flake8 errors.

jnothman commented on sklearn/semi_supervised/label_propagation.py:
```diff
@@ -264,12 +259,15 @@ def fit(self, X, y):
-               and remaining_iter > 1):
+
+        self.n_iter_ = 0
+        while self.n_iter_ < self.max_iter:
```
I suppose we could now do this as a for loop...

musically-ut replied:
You mean setting `self.n_iter_` using the loop variable?
I personally feel a bit uncomfortable using the loop variable outside the loop. 

(* angsty feeling *)

musically-ut replied:
So ... shall I make this change to expedite the merge?

amueller replied:
I think he meant
```python
for n_iter in range(self.max_iter):
```

musically-ut replied:
With the `for` loop, there are two alternatives:

## Alternative 1

Setting `self.n_iter_` after the `for` loop. It would look a bit ugly because we'll need to figure out whether in the last iteration the loop the tolerance condition was met (then `self.n_iter_ = n_iter`) or the `else:` branch was reached (in which case `self.n_iter_ = n_iter + 1`):

```python
self.n_iter_ = 0
for n_iter in range(self.max_iter):
    if converged:
        break
    # ...
else:
    # warn
    self.n_iter_ = 1 # Count the last iteration.

self.n_iter_ += n_iter
```

## Alternate 2

Leaving `self.n_iter_ += 1` inside the for-loop. In this case case the loop variable (i.e. `n_iter`) is not used anywhere.

```python
self.n_iter_ = 0
for n_iter in range(self.max_iter):
    if converged:
        break
    # ...
    self.n_iter_ += 1
else:
    # warn
```

----

Were any of these versions what you (both of you) had in mind? Personally, I like the `while` loop and then Alternative 2. :)

amueller replied:
why not use Alternative 2 and set ``self.n_iter_ = n_iter`` after the loop?

I don't understand the ``self.n_iter_ = 1 # Count the last iteration.`` line.

musically-ut replied:
@amueller I think that you meant Alternative 1 with `self.n_iter_ = n_iter` after the loop?

If so, then the problem happens on the last iteration: `self.n_iter_` will never be equal to `max_iter` with your suggestion.

amueller replied:
Sorry I still don't follow.
I meant use
```python
for n_iter in range(self.max_iter):
    if converged:
        break
    # ...
else:
    # warn
self.n_iter_ = n_iter
```
I guess we're counting from 1 here, so maybe 
```python
self.n_iter_ = n_iter + 1
```

musically-ut replied:
@amueller In the new proposal, `self.n_iter_`  can never be zero, and zero is a valid value technically. The handling of these corner cases is what makes the for loop slightly ugly.

I'm sorry, I am away from my laptop and am a bit constrained in my explanations.

jnothman replied:
This comment was a cosmetic one that is not worth withholding merge for.

musically-ut replied:
So ... is that a "_go_" for the `while` loop? :-)

jnothman replied:
Whatever you think most readable. I'd prefer

```
for self.n_iter_ in ...:
   ...
else:
   ...
```

I think, but it matters very little

musically-ut replied:
First, I didn't know that we could use `self.n_iter_` in the `for` loop; today I learned. :)

Second, this is the best implementation I can think of which has the same behavior as the original `while` loop:

```python
for self.n_iter_ in range(max_iter):
    # ...
else:
    # ...
    self.n_iter_ += 1
```

The `self.n_iter_ += 1` in the `else:` clause is necessary to ensure that `self.n_iter_ == self.max_iter` is possible and that it correctly happens if the loop doesn't `break` out due to convergence. This was the issue we were out to fix originally. :)

With this implementation, I am quite happy with the `for` loop as well. I'll make this change. :+1:

jnothman reviewed (commented):
LGTM

amueller commented:
Do we want this for 0.19?

amueller commented:
LGTM, no strong opinion on the for loop thought it would be slightly nicer.

jnothman commented:
I'm happy for this to be merged. I don't consider it essential for 0.19, but we are newly encouraging people to use these estimators, so why not.

musically-ut committed:
While loop -> for loop.

Also, add tests for verify that n_iter_ == max_iter if warning is
raised.

musically-ut commented:
Bump!
